extensions/cornerstone-dynamic-volume/src/panels/PanelGenerateImage.tsx

```diff
@@ -0,0 +1,414 @@
+  const [timeOptions, setTimeOptions] = useState([]);
+  const [rangeValues, setRangeValues] = useState([]);
+  const [sliderValues, setSliderValues] = useState([]);
+  const [timeFramesToUse, setTimeFramesToUse] = useState([]);
+  const [computedDisplaySet, setComputedDisplaySet] = useState(null);
+  const [myDynamicVolume, setMyDynamicVolume] = useState(null);
+  const [myComputedVolume, setMyComputedVolume] = useState(null);
+  const [frameRate, setFrameRate] = useState(20);
+  const [uuidComputedVolume, setUuidComputedVolume] = useState(
+    csUtils.uuidv4()
+  );
+
+  // Cine states
+  // cineState is passed from the reducer in cine provider, cineService is the
+  // api passed from the reducer
+  const [cineState, cineService] = useCine();
+  const { inCineEnabled, cines } = cineState;
+
+  const handleGenerateOptionsChange = generateOptions => {
+    setGenerateOptions(prevState => {
+      const newState = { ...prevState };
+      Object.keys(generateOptions).forEach(key => {
+        if (typeof generateOptions[key] === 'object') {
+          newState[key] = {
+            ...prevState[key],
+            ...generateOptions[key],
+          };
+        } else {
+          newState[key] = generateOptions[key];
+        }
+      });
+      return newState;
+    });
+  };
+
+  // Establish a reference to the viewer API context
+  const { activeViewportIndex, viewports } = viewportGridService.getState();
+  const displaySetInstanceUID = viewports[0].displaySetInstanceUIDs[0];
+  const volumeLoaderScheme = 'cornerstoneStreamingDynamicImageVolume'; // Loader id which defines which volume loader to use
+  const computedVolumeId = `cornerstoneStreamingImageVolume:${uuidComputedVolume}`;
+  const dynamicVolumeId = `${volumeLoaderScheme}:${displaySetInstanceUID}`;
+
+  useEffect(() => {
+    // ~~ Subscription
+    const added = cornerstoneViewportService.EVENTS.VIEWPORT_DATA_CHANGED;
+    const subscriptions = [];
+
+    // Used to make sure initialization of computed volume only runs once
+    let initStatus = false;
+
+    [added].forEach(evt => {
+      subscriptions.push(
+        cornerstoneViewportService.subscribe(evt, evtdetails => {
+          evtdetails.viewportData.data.forEach(volumeData => {
+            if (volumeData.volumeId.split(':')[0] === volumeLoaderScheme) {
+              if (!initStatus) {
+                initStatus = true;
+                setMyDynamicVolume(volumeData.volume);
+                const options = numTimePointsToOptions(
+                  volumeData.volume._numTimePoints
+                );
+                const range = [1, volumeData.volume._numTimePoints];
+                setTimeOptions(prevArray => [...prevArray, ...options]);
+                setRangeValues(prevArray => [...prevArray, ...range]);
+                setSliderValues(range);
+                setTimeFramesToUse([0, 1]);
+                const computedVolumeInit = createComputedVolume(
```

Comment:
shouldn't do on init, should be on button click

Reply:
Moved to onGenerateImage